### PR TITLE
Pre-trained data preprocess

### PR DESCRIPTION
**Pull request by Steffy-zxf**
### PR types

 New features

### PR changes

Models

### Describe

* add pretrained data process
* support bert  basic token masking, whole word masking and  entity masking strategy
* create pretraining data according to model name

### CODE REVIEW

File: examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import glob
+import os
+
+from datasets import load_dataset
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py:18]: 文档是否需要突出新增HF datasets的依赖?

File: examples/language_model/pretraining_data_prepare/text_sharding.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+import statistics
+from collections import defaultdict
+from itertools import islice
+
+import nltk
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/text_sharding.py:20]: 检查下依赖项requirements.txt

File: examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class BookscorpusTextFormatter:
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py:21]: 这种类型是固定数据类型，如wiki和bookcorpus。
需要额外考虑如果用户自定义数据集，应该按什么形式的各式进行准备会更加通用。
我理解wiki那种形式数据不是通用形式。

**Review by ZeyuChen — COMMENTED** [review]
Check requirements and consider about customized dataset from user. How to deal with the format.

File: examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py (hunk 1 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py:18]: 需要的 已增加requirements.txt文件。文档待补充。

File: examples/language_model/pretraining_data_prepare/text_sharding.py (hunk 2 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/text_sharding.py:20]: 同上

File: examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py (hunk 3 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/text_formatting/bookcorpus.py:21]: 是的。但是数据预处理我们没办法控制，只能交给用户自己处理。由于中英文维基和BookCorpus是模型预训练常用的数据，所以这里内置了常用的数据集预处理过程。

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+"""Create masked LM/next sentence masked_lm examples for BERT/BERT-wwm/ERNIE 1.0."""
+
+import argparse
+import collections
+import h5py
+import random
+import re
+
+import numpy as np
+from LAC import LAC
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py:25]: 这里引入LAC是否确认过不存在paddle 2.0版本的兼容问题？

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+from paddlenlp.utils.log import logger
+from paddlenlp.transformers import BertTokenizer, ErnieTokenizer
+from tqdm import tqdm
+
+# chinese word segmentation tool
+lac_cws = LAC(mode='seg')
+# chinese lexical analysis tool
+lac = LAC(mode="lac")
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py:33]: 是否还有必要开lac模式呢？是否默认seg模式作为lac即可。

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+        s += "\n"
+        return s
+
+    def __repr__(self):
+        return self.__str__()
+
+
+def write_instance_to_example_file(instances, tokenizer, max_seq_length,
+                                   max_predictions_per_seq, output_file):
+    """Create TF example files from `TrainingInstance`s."""
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: 这里的注释不正确，不应该是TF example

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+
+    return entity_words
+
+
+def get_whole_word_mask_tokens(tokens,
+                               max_word_length=4,
+                               chinese_word_segmentation_fn=None,
+                               entity_words=None):
+    """
+    Do whole word mask on chinese word.
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: `chinese word` -> `Chinese words`

Comment (ZeyuChen) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: Chinese should begin with capital letter in comments.

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 8]
```diff
@@ -0,0 +1,672 @@
...
+                    words[index] = entity_word
+    else:
+        words = chinese_word_segmentation_fn("".join(tokens))
+
+    words_dict = {x: 1 for x in words}
+
+    new_tokens = []
+    i = 0
+    while i < len(tokens):
+
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: remove useless blank line.

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 9]
```diff
@@ -0,0 +1,672 @@
...
+        # We want to sometimes truncate from the front and sometimes from the
+        # back to add more randomness and avoid biases.
+        if rng.random() < 0.5:
+            del trunc_tokens[0]
+        else:
+            trunc_tokens.pop()
+
+
+def main():
+
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: remove useless blank line

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 5 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py:33]: 开启lac模式，是为了识别实体，用于entity masking

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 4 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py:25]: 确认，LAC最新版本兼容paddle2.0版本

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 6 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: 已修改

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 7 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: done

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 8 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: done

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 9 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py]: done

File: examples/language_model/pretraining_data_prepare/README.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
+# Transformers模型继续预训练
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/README.md]: 标题改为 预训练数据准备 好像更恰当

File: examples/language_model/pretraining_data_prepare/README.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+nltk
+LAC
+tqdm
+```
+
+通过`pip install -r requirements.txt`即可完成依赖安装。
+
+## 快速开始
+
+### 清洗原数据
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/README.md]: 清洗原数据 -> 数据清洗

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 12]
```diff
@@ -0,0 +1,635 @@
...
+
+def get_whole_word_mask_tokens(tokens, words, max_word_length=4):
+    """
+    Do whole word mask on Chinese word.
+    First, we do Chinese word segmentation on the sequence of tokens, which are from the WordPiece tokenization.
+    Then, we add the '##' mark on chinese characters which are in the middle of Chinese words.
+    And if the tokens are not chinese characters, we just exploit the results of WordPiece tokenization as words.
+    Such as, 
+         - text line : 通过利用mercer核，将样本从输入空间映射到高维特征空间，使原来没有显现的特征突现出来，取得了很好的图像分割效果。
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py:236]: `取得了很好的图像分割效果` ? 为什么是图像分割？

File: examples/language_model/pretraining_data_prepare/README.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+首先，我们需要将原始数据进行清洗，清洗成一行一篇文章的形式，并且文章之间以空行进行分隔。
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/README.md]: 听这句话的意思，感觉这一过程叫清洗是否不够准确，感觉是Normalization或者事Formatting 更加准确

Comment (ZeyuChen) [reply examples/language_model/pretraining_data_prepare/README.md]: 因为清洗的意思更多是filter掉脏数据或者脏词

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py [hunk 14]
```diff
@@ -0,0 +1,635 @@
...
+    """Save instance data to HDF5 file."""
+    total_written = 0
+    features = collections.OrderedDict()
+
+    num_instances = len(instances)
+    features["input_ids"] = np.zeros(
+        [num_instances, max_seq_length], dtype="int32")
+    features["input_mask"] = np.zeros(
+        [num_instances, max_seq_length], dtype="int32")
+    features["segment_ids"] = np.zeros(
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/create_pretraining_data.py:77]: 对于这里的定义采用segment_ids, 会不会与我们模型底层用token_type_ids显得不是一个体系的，不够匹配？

File: examples/language_model/pretraining_data_prepare/README.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+```text
+pretraining_data_prepare/
+├── create_pretraining_data.py # MLM数据构建
+├── main.py # 程序入口文件
+├── README.md # 文档说明
+├── requirements.txt # 第三方依赖
+├── text_formatting # 数据清洗
+│   ├── bookcorpus.py # bookcorpus数据清洗脚本
+│   └── wikicorpus.py # 中英文维基百科数据清洗集哦啊笨
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/README.md]: 洗集哦啊笨 -> 脚本

File: examples/language_model/pretraining_data_prepare/README.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+以下是本项目主要代码结构及说明：
+
+```text
+pretraining_data_prepare/
+├── create_pretraining_data.py # MLM数据构建
+├── main.py # 程序入口文件
+├── README.md # 文档说明
+├── requirements.txt # 第三方依赖
+├── text_formatting # 数据清洗
```
Comment (ZeyuChen) [inline examples/language_model/pretraining_data_prepare/README.md]: 感觉叫格式化更位准确。

**Review by ZeyuChen — COMMENTED** [review]
1. 术语可能不够准确，formatting和清洗不是一个概念。
2. 底层segment_ids和我们模型自身token_type_ids这地方是否需要统一？

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 12 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py:236]: 这是举例说明。

File: examples/language_model/pretraining_data_prepare/create_pretraining_data.py (hunk 14 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/create_pretraining_data.py:77]: 由于bert预训练脚本中也是使用"segment_ids"，所以这里也用的是"segment_ids"。

File: examples/language_model/pretraining_data_prepare/README.md (hunk 10 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/README.md]: Done

File: examples/language_model/pretraining_data_prepare/README.md (hunk 11 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/README.md]: 已修改为 数据格式化

File: examples/language_model/pretraining_data_prepare/README.md (hunk 13 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/README.md]: 数据清洗 -> 数据格式化

File: examples/language_model/pretraining_data_prepare/README.md (hunk 15 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/README.md]: 已修改

File: examples/language_model/pretraining_data_prepare/README.md (hunk 16 above)
Comment (Steffy-zxf) [reply examples/language_model/pretraining_data_prepare/README.md]: done

**Review by ZeyuChen — APPROVED** [review]
LGTM
